File: Practica02/ejercicio2/fraccion.py
"""
La Clase Fraccion .
ahora franz
1.0 03/03/2026
"""
class Fraccion :
	def __init__(self, a, b):
		self.__numerador = a
		self.__denominador = b

	def __add__(self, o):
		a = self.__numerador * o.__denominador + \
			self.__denominador * o.__numerador
		b = self.__denominador * o.__denominador
		return Fraccion(a, b)

	def __sub__(self, o):
		c = Fraccion(0, 1)
		c.__numerador = self.__numerador * o.__denominador - \
			self.__denominador * o.__numerador
		c.__denominador = self.__denominador * o.__denominador
		return c

	def __str__(self):
		return " {}/{} ".format(self.__numerador, self.__denominador)

	# Inciso 1: Multiplica dos fracciones usando el operador *
	def __mul__(self, o):
		a = self.__numerador * o.__numerador
		b = self.__denominador * o.__denominador
		return Fraccion(a, b)

	# Inciso 2: Divide dos fracciones usando el operador /
	def __truediv__(self, o):
		a = self.__numerador * o.__denominador
		b = self.__denominador * o.__numerador
		return Fraccion(a, b)

	# Inciso 3: Compara si dos fracciones son iguales usando ==
	def __eq__(self, o):
		if not isinstance(o, Fraccion):
			return False
		return self.__numerador * o.__denominador == self.__denominador * o.__numerador

	# Inciso 4: Convierte la fracción a decimal validando división entre cero
	def convertirADecimal(self):
		if self.__denominador == 0:
			print("Error: División entre cero")
			return None
		return self.__numerador / self.__denominador

	# Inciso 5: Comprueba si esta fracción es inversa de otra usando el operador *
	def esInverso(self, o):
		producto = self * o
		return producto.convertirADecimal() == 1.0

	# Inciso 6: Convierte una cadena "a/b" en un objeto Fraccion
	@staticmethod
	def parseFraccion(cadena):
		partes = cadena.strip().split('/')
		n = int(partes[0])
		d = int(partes[1])
		return Fraccion(n, d)

	# Inciso 7: Retorna una nueva Fraccion simplificada
	def simplifica(self):
		import math
		mcd = math.gcd(self.__numerador, self.__denominador)
		return Fraccion(self.__numerador // mcd, self.__denominador // mcd)

File: Practica02/ejercicio1/Fraccion.java
package ejercicio1;

/**
* La Clase Fracci´on .
*
* @author Franz xd
* @version 1.0 03/03/2026
*
*/
public class Fraccion {
	// Atributos
	private int numerador;
	private int denominador;

	// M´etodos
	/**
	 * Construye una nueva fracci´on .
	 */
	public Fraccion() {
		this.numerador = 0;
		this.denominador = 1;
	}

	/**
	 * Construye una nueva fracci´on .
	 */
	public Fraccion(int n, int d) {
		this.numerador = n;
		this.denominador = d;
	}

	/**
	 * Comprueba si dos fracciones son iguales .
	 *
	 * @param o Object , otro objeto .
	 * @return true si las fracciones son iguales .
	 */
	@Override
	public boolean equals(Object o) {
		if (o instanceof Fraccion) {
			Fraccion f = (Fraccion) o;
			return f.numerador == this.numerador && f.denominador == this.denominador;
		} else
			return false;
	}

	/**
	 * Retorna la representaci´on de Cadena del objeto Fraccion .
	 *
	 * @return String , la fracci´on en formato de cadena .
	 */
	@Override
	public String toString() {
		return String.format(" %d/ %d", this.numerador, this.denominador);
	}

	/**
	 * Suma dos fracciones .
	 *
	 * @param o Fraccion , contiene la otra fraccion .
	 * @return Fraccion resultante .
	 */
	public Fraccion suma(Fraccion o) {
		int n = (this.numerador * o.denominador) + (this.denominador * o.numerador);
		int d = this.denominador * o.denominador;
		return new Fraccion(n, d);
	}

	/**
	 * Resta dos fracciones .
	 *
	 * @param o Fraccion , contiene la otra fracci´on .
	 * @return c Fraccion resultante .
	 */
	public Fraccion resta(Fraccion o) {
		Fraccion c = new Fraccion();
		c.numerador = (this.numerador * o.denominador) - (this.denominador * o.numerador);
		c.denominador = this.denominador * o.denominador;
		return c;
	}
    // Inciso a: Multiplica dos fracciones
    public Fraccion multiplica(Fraccion o) {
        int n = this.numerador * o.numerador;
        int d = this.denominador * o.denominador;
        return new Fraccion(n, d);
    }

    // Inciso b: Divide dos fracciones
    public Fraccion divide(Fraccion o) {
        int n = this.numerador * o.denominador;
        int d = this.denominador * o.numerador;
        return new Fraccion(n, d);
    }

    // Inciso c: Convierte la fracción a número decimal validando división entre 0
    public double convertirADecimal() {
        if (this.denominador == 0) {
            System.out.println("Error: No se puede dividir entre cero.");
            return 0.0;
        }
        return (double) this.numerador / this.denominador;
    }

    // Inciso d: Verifica si esta fracción es la inversa de otra usando el método multiplica()
    public boolean esInverso(Fraccion o) {
        Fraccion producto = this.multiplica(o);
        return producto.convertirADecimal() == 1.0;
    }
    // Inciso e: Convierte una cadena con formato "a/b" en un objeto Fraccion
    public static Fraccion parseFraccion(String str) {
        String[] partes = str.trim().split("/");
        int n = Integer.parseInt(partes[0].trim());
        int d = Integer.parseInt(partes[1].trim());
        return new Fraccion(n, d);
    }

    // Método auxiliar para calcular el Máximo Común Divisor (MCD)
    private int mcd(int a, int b) {
        a = Math.abs(a);
        b = Math.abs(b);
        while (b != 0) {
            int temp = b;
            b = a % b;
            a = temp;
        }
        return a;
    }

    // Inciso f: Retorna una nueva Fraccion simplificada hasta su forma irreducible
    public Fraccion simplifica() {
        int divisor = mcd(this.numerador, this.denominador);
        if (divisor == 0) {
            return new Fraccion(this.numerador, this.denominador);
        }
        return new Fraccion(this.numerador / divisor, this.denominador / divisor);
    }
}
